drivers/stepper/step_dir_stepper_common.c

```diff
@@ -0,0 +1,263 @@
+		ret = gpio_pin_set_dt(&config->dir_pin, 1);
+		break;
+	case STEPPER_DIRECTION_NEGATIVE:
+		ret = gpio_pin_set_dt(&config->dir_pin, 0);
+		break;
+	default:
+		LOG_ERR("Unsupported direction");
+		return -ENOTSUP;
+	}
+	if (ret < 0) {
+		LOG_ERR("Failed to set direction");
+		return ret;
+	}
+
+	if (config->dual_edge) {
+		ret = gpio_pin_toggle_dt(&config->step_pin);
+	} else {
+		ret = gpio_pin_toggle_dt(&config->step_pin);
+		ret |= gpio_pin_toggle_dt(&config->step_pin);
+	}
+
+	if (ret < 0) {
+		LOG_ERR("Failed to toggle step pin");
+		return ret;
+	}
+
+	return 0;
+}
+
+static void update_remaining_steps(struct step_dir_stepper_common_data *data)
+{
+	if (data->step_count > 0) {
+		data->step_count--;
+		(void)k_work_reschedule(&data->stepper_dwork, K_USEC(data->delay_in_us));
+	} else if (data->step_count < 0) {
+		data->step_count++;
+		(void)k_work_reschedule(&data->stepper_dwork, K_USEC(data->delay_in_us));
+	} else {
+		if (!data->callback) {
+			LOG_WRN_ONCE("No callback set");
+			return;
+		}
+		data->callback(data->dev, STEPPER_EVENT_STEPS_COMPLETED, data->event_cb_user_data);
+	}
+}
+
+static void update_direction_from_step_count(const struct device *dev)
+{
+	struct step_dir_stepper_common_data *data = dev->data;
+
+	if (data->step_count > 0) {
+		data->direction = STEPPER_DIRECTION_POSITIVE;
+	} else if (data->step_count < 0) {
+		data->direction = STEPPER_DIRECTION_NEGATIVE;
+	} else {
+		LOG_ERR("Step count is zero");
+	}
+}
+
+static void position_mode_task(const struct device *dev)
+{
+	struct step_dir_stepper_common_data *data = dev->data;
+
+	if (data->step_count) {
+		(void)step_dir_stepper_perform_step(dev);
+	}
+	update_remaining_steps(dev->data);
+}
+
+static void velocity_mode_task(const struct device *dev)
+{
+	struct step_dir_stepper_common_data *data = dev->data;
+
+	(void)step_dir_stepper_perform_step(dev);
+	(void)k_work_reschedule(&data->stepper_dwork, K_USEC(data->delay_in_us));
+}
+
+static void stepper_work_step_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct step_dir_stepper_common_data *data =
+		CONTAINER_OF(dwork, struct step_dir_stepper_common_data, stepper_dwork);
+
+	K_SPINLOCK(&data->lock) {
+		switch (data->run_mode) {
+		case STEPPER_RUN_MODE_POSITION:
+			position_mode_task(data->dev);
+			break;
+		case STEPPER_RUN_MODE_VELOCITY:
+			velocity_mode_task(data->dev);
+			break;
+		default:
+			LOG_WRN("Unsupported run mode %d", data->run_mode);
+			break;
+		}
+	}
+}
+
+int step_dir_stepper_common_init(const struct device *dev)
+{
+	const struct step_dir_stepper_common_config *config = dev->config;
+	struct step_dir_stepper_common_data *data = dev->data;
+	int ret;
+
+	if (!gpio_is_ready_dt(&config->step_pin) || !gpio_is_ready_dt(&config->dir_pin)) {
+		LOG_ERR("GPIO pins are not ready");
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&config->step_pin, GPIO_OUTPUT);
+	if (ret < 0) {
+		LOG_ERR("Failed to configure step pin");
```

Comment:
I'd suggest adding the return code on all LOG_ERR messages

Reply:
Done